dask_jobqueue/remote_slurm.py

```diff
@@ -0,0 +1,174 @@
+
+logger = logging.getLogger(__name__)
+
+
+class RemoteSLURMJob(SLURMJob):
+    def __init__(self, *args, **kwargs):
+        self.api_client_session = kwargs.pop("api_client_session")
+        self.api_url = kwargs.pop("api_url")
+        self.remote_job_extra = (
+            kwargs.pop("remote_job_extra") if "remote_job_extra" in kwargs else {}
+        )
+        super().__init__(*args, **kwargs)
+
+    @contextmanager
+    def job_file(self):
+        """
+        Override the job_file and pass back the job_script. job_file() is used by Job.start and
+         gives output directly to SLURMJob._submit_job().
+        """
+        yield self.job_script()
+
+    @property
+    def _job_configuration(self) -> Dict:
+        configuration = {
+            "environment": {"name": "test"},
+            "tasks": 1,
+            "cpus_per_task": self.job_cpu or self.worker_cores,
+        }
+
+        if self.job_name is not None:
+            configuration["name"] = self.job_name
+        if self.log_directory is not None:
+            configuration[
+                "standard_error"
+            ] = f"{self.log_directory}/{self.job_name or 'worker'}-%%J.err"
+            configuration[
+                "standard_output"
+            ] = f"{self.log_directory}/{self.job_name or 'worker'}-%%J.out"
+        if self.queue is not None:
+            configuration["partition"] = self.queue
+        if self.project is not None:
+            configuration["account"] = self.project
+
+        memory = self.job_mem
+        if self.job_mem is None:
+            memory = slurm_format_bytes_ceil(self.worker_memory)
+        if memory is not None:
+            configuration["memory_per_node"] = memory
+
+        if self.walltime is not None:
+            configuration["time_limit"] = self.walltime
+
+        configuration.update(self.remote_job_extra)
+
+        return configuration
+
+    async def _submit_job(self, script):
+        # https://slurm.schedmd.com/rest_api.html#slurmctldSubmitJob
+        try:
+            response = await self.api_client_session.post(
+                f"{self.api_url}slurm/v0.0.36/job/submit",
+                json={"script": script, "job": self._job_configuration},
+            )
+            response.raise_for_status()
+        except aiohttp.ClientError as e:
+            logger.exception("SLURMJob request failed.")
+            raise RuntimeError from e
+        else:
+            return await response.json()
+
+    def _job_id_from_submit_output(self, out):
+        # out is the JSON output from _submit_job request.post
+        # See https://slurm.schedmd.com/rest_api.html#v0.0.36_job_submission_response
+        return out["job_id"]
+
+    async def close(self):
+        logger.debug("Stopping worker: %s job: %s", self.name, self.job_id)
+        # https://slurm.schedmd.com/rest_api.html#slurmctldCancelJob
+        try:
+            response = await self.api_client_session.delete(
```

Comment:
@jacobtomlinson I'm not super familiar with aiohttp. Now when I run pytest it passes succesfully (after the last test) but then prints;

```
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x7fc5a2477e10>
Unclosed connector
connections: ['[(<aiohttp.client_proto.ResponseHandler object at 0x7fc5a20f25f8>, 1007195.255574382)]']
connector: <aiohttp.connector.UnixConnector object at 0x7fc5a2477898>
Task was destroyed but it is pending!
task: <Task pending coro=<AdaptiveCore.adapt() done, defined at /opt/anaconda/lib/python3.6/site-packages/distributed/deploy/adaptive_core.py:179> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7fc5a24e0c48>()]> cb=[IOLoop.add_future.<locals>.<lambda>() at /opt/anaconda/lib/python3.6/site-packages/tornado/ioloop.py:688]>
```

Any ideas how to terminate the session properly when it's being passed in like this?

Reply:
Figured it out :)